Return NIL when FIND-PLACE is called with non-existent name

mathrick opened the pull request:
This makes a bit more sense than returning the entire pipeline, which was the case previously but seems to be an unintential side-effect of `NIL` returned from `GETHASH` also being a valid list and thus path

mathrick committed:
Return NIL when FIND-PLACE is called with non-existent name

This makes a bit more sense than returning the entire pipeline, which
was the case previously but seems to be an unintential side-effect of
NIL returned from GETHASH also being a valid list and thus path

Shinmera commented on pipeline.lisp:
```diff
@@ -36,7 +36,10 @@ As such, (1 4) matches the following:
+    (multiple-value-bind (path found)
+        (gethash name (names pipeline))
+      (when found
+        (find-place (pipeline pipeline) path)))))
```
It's been a while since I looked at this library, but can't it just be `(or (gethash name (names pipeline)) (find-place (pipeline pipeline) path))`?

mathrick replied:
Not entirely sure what you mean. It seems that what you wrote would *either* return the path (ie. the sequence of ints) if `NAME` was found in `NAMES`, *or* it would try to use an unbound variable (since it no longer binds `PATH` presumably).

mathrick commented:
Actually, this patch seems to cause failures in `ADD-SEGMENT`, I'm working on a fix.

mathrick commented:
OK, that should work better. I had no idea that `NIL` is a symbol and not a list when it comes to method selection.

mathrick committed:
Add RESOLVE-PLACE and use it in FIND-PLACE when a name is passed

An unknown name now signals an error by default.

mathrick committed:
Accept named places in all high-level pipeline operations

Whenever a pipeline argument is accepted, place can now be a symbolic
name. Since names are only meaningful in the context of a pipeline
that stores them, low-level operations which accept an array directly
remain unable to accept names.

mathrick commented:
I've decided to rework my approach, as it made more sense to signal an error instead when an unknown name is passed in. This makes the API reasonable for subsequent commits that I've added that also fix #1.

Summary of changes:
- New generic function called `RESOLVE-PLACE` has been added, which handles resolving a name to a full path. All other code delegates to it for name resolution. By default, it signals an error if an unknown name is given, but can optionally be told to ignore unknown names and not signal. In this case, secondary value can be used to check if the name was successfuly resolved. `FIND-PLACE` no longer returns a secondary value with this change
- All functions which accept a pipeline object and a place now allow the use of names. This makes the API consistent, and also fixes some operations which were extremely cumbersome previously, such as `REMOVE-SEGMENT` mentioned in #1
- Since the name resolution is done via `RESOLVE-PLACE` in the default mode, all those functions will also signal an error if given an unknown name. This seems like the correct approach, since it avoids weird consequences of things sometimes resolving to `NIL`, which can give unexpected results and makes the API hard to design and reason about. User code can still test whether names are known and what they resolve to by using `RESOLVE-PLACE` explicitly

Please let me know if this looks like a good approach to you, and whether the PR itself is the way you'd like to have it. If you need me to split things up or something to make it easier for you, I can do that. Note that I did not re-generate documentation, since I don't know how you update `gh-pages`, but that should probably be done before it gets merged in.

Shinmera commented on pipeline.lisp:
```diff
@@ -15,28 +15,64 @@
    (%names :initarg :names :initform (make-hash-table :test 'equal) :accessor names))
   (:documentation "Base pipeline object that is necessary to orchestrate piping systems."))
 
+(define-condition unknown-name-error ()
```
condition needs to be exported

Shinmera commented on pipeline.lisp:
```diff
@@ -15,28 +15,64 @@
+  (:report (lambda (c stream)
+             (format stream "Name ~A is not known in pipeline ~A" (name c) (pipeline c)))))
+
+(defgeneric resolve-place (pipeline place &key ignore-unknown)
```
It is more standard to have the parameter be `if-does-not-exist` and the permissible values be: `NIL` (return NIL) and `:error:` (return error)

Shinmera commented on pipeline.lisp:
```diff
@@ -15,28 +15,64 @@
+  Returns a secondary value which is T if the place was successfully resolved, and NIL if
+  it was an unknown name and INGORE-UNKNOWN was specified.")
+  (:method ((pipeline pipeline) (place list) &key &allow-other-keys)
+    (values place t))
```
`t` and `nil` are constants and should thus be uppercase: `T` `NIL`.

mathrick replied:
You mean in the code? Cause they already are in the docstring. I've never seen anyone do it that way in the code, and in fact that would be harmful for if the readtable-case is `:invert`, [which is very useful](https://www.cliki.net/case%20sensitivity) and I've been trying to get libraries usable with it for a long time.

Shinmera replied:
Yes in the code. This is how it is in all my code and libraries, and it will continue to be so.

Shinmera replied:
Besides, anyone changing the readtable case or other kinds of funny business for libraries they don't control should expect just as much mayhem as if they set `*read-base*` to 8 or anything similar.

mathrick replied:
It's only "funny business" because people assume `:UPCASE`, thus making anything else unusable, but sure. Your code already follows that.

Shinmera commented on pipeline.lisp:
```diff
@@ -15,28 +15,64 @@
+    (values place t))
+  (:method ((pipeline pipeline) (place symbol) &key ignore-unknown)
+    (multiple-value-bind (path found)
+        (gethash place (names pipeline))
```
It isn't valid to store a NIL path as a named path, so using the `found` complication here seems unnecessary to me.

mathrick replied:
OK, if `NIL` is specifically invalid, then I can certainly simplify this.

Shinmera commented on pipeline.lisp:
```diff
@@ -46,17 +82,20 @@ As secondary value it returns the position of the matched item within the parent
     (error "Cannot descend into segment!"))
   (:method ((array array) (place list))
     (if (<= (length place) 1)
+
```
What is this stray newline?

Shinmera commented on pipeline.lisp:
```diff
@@ -83,21 +122,26 @@ segments. You should always use remove-segment instead if possible.")
-If place is set, the pipe is added to the specified place as per INSERT.
-The place specified is expected to be a splitter or similar to append
-the pipe to. Returns the segment.")
+  (:documentation "Add a new segment to the pipeline.  If place is set, the pipe is added
```
Why was this reformatted? The first line of a docstring should always be a single-sentence short description.

mathrick replied:
Likely because I went back and forth on what the API should be, and then I missed that the docstring got reformatted even though I reverted the changes to the code.

Shinmera reviewed (changes requested):
Mostly stylistic stuff.

Shinmera commented on pipeline.lisp:
```diff
@@ -15,28 +15,64 @@
   (:method ((pipeline pipeline) (name symbol))
-    (find-place (pipeline pipeline) (gethash name (names pipeline)))))
+    (if name
+        (find-place pipeline (resolve-place pipeline name))
```
Would be nice to provide a `use-value` restart here.

mathrick replied:
Makes sense
